test-avocado/lib/libjournal.py

```diff
@@ -0,0 +1,151 @@
+# -*- coding: utf-8 -*-
+
+# This file is part of Cockpit.
+#
+# Copyright (C) 2013 Red Hat, Inc.
+#
+# Cockpit is free software; you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation; either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# Cockpit is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Tools for writing Cockpit test cases.
+"""
+
+import re       # journal matching
+import logging  # debug output in classes
+from avocado.utils import process # Journal output
+
+log=logging.getLogger("Journal")
+
+admins_only_pam = """
+#%PAM-1.0
+auth       required     pam_sepermit.so
+auth       substack     password-auth
+auth       include      postlogin
+auth       optional     pam_reauthorize.so prepare
+account    required     pam_nologin.so
+account    sufficient   pam_succeed_if.so uid = 0
+account    required     pam_succeed_if.so user ingroup wheel
+account    include      password-auth
+password   include      password-auth
+# pam_selinux.so close should be the first session rule
+session    required     pam_selinux.so close
+session    required     pam_loginuid.so
+# pam_selinux.so open should only be followed by sessions to be executed in the user context
+session    required     pam_selinux.so open env_params
+session    optional     pam_keyinit.so force revoke
+session    optional     pam_reauthorize.so prepare
+session    include      password-auth
+session    include      postlogin
+"""
+
+class Journalctl:
+    def __init__(self):
+        pass
+
+    def journal_messages(self, syslog_ids, log_level):
+        """Return interesting journal messages"""
+
+        # Journald does not always set trusted fields like
+        # _SYSTEMD_UNIT or _EXE correctly for the last few messages of
+        # a dying process, so we filter by the untrusted but reliable
+        # SYSLOG_IDENTIFIER instead
+
+        matches = " ".join(map(lambda id: "SYSLOG_IDENTIFIER=" + id, syslog_ids))
+
+        # Some versions of journalctl terminate unsuccessfully when
+        # the output is empty.  We work around this by ignoring the
+        # exit status and including error messages from journalctl
+        # itself in the returned messages.
+
+        cmd = "journalctl 2>&1 -o cat -p %d %s" % (log_level, matches)
+        out=process.run(cmd, shell=True, ignore_status=True)
```

Comment:
Added spaces.  The ignore_status is justified to work around a journalctl oddness as explained in the comment.